File: packages/client-python/src/rocketride/mixins/connection.py
# MIT License
#
# Copyright (c) 2026 Aparavi Software AG
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

"""
Connection Management for RocketRide Client.

This module handles connecting to and disconnecting from RocketRide servers.
It manages the WebSocket connection lifecycle, authentication, and status tracking.

The connection system automatically handles:
- WebSocket connection establishment
- Authentication with your credential (API key, Zitadel access_token, or rr_ user token)
- Connection status tracking
- Automatic reconnection on disconnects (when persist=True)
- Graceful disconnection and cleanup

Usage:
    client = RocketRideClient(uri="http://localhost:8080")
    result = await client.connect("your_api_key")
    # result is a ConnectResult with full user identity and organizations

    if client.is_connected():
        # Do work with the client
        pass

    await client.disconnect()
"""

# Design: Physical connect/disconnect live in _internal_connect and _internal_disconnect.
# Non-persist mode: connect() and disconnect() call those directly.
# Persist mode: connect() uses _attempt_connection (first attempt inline); retries and
# reconnect-on-disconnect are scheduled via _schedule_reconnect / _attempt_reconnect.

import asyncio
import os
import time
import urllib.parse
from typing import Any, Dict, Optional
from ..core import DAPClient, TransportWebSocket, CONST_DEFAULT_WEB_PORT, CONST_DEFAULT_WEB_PROTOCOL
from ..core.exceptions import AuthenticationException
from ..types.client import ConnectResult


class ConnectionMixin(DAPClient):
    """
    Handles connection and disconnection to RocketRide servers.

    This mixin provides the fundamental connection management capabilities
    for the RocketRide client. It manages WebSocket connections, handles
    authentication, and tracks connection status.

    Key Features:
    - Establishes secure WebSocket connections to RocketRide servers
    - Single connect(credential) call authenticates and returns ConnectResult
    - Tracks connection status for reliable operations
    - Automatic reconnection on disconnect (when persist=True)
    - Provides graceful connection cleanup
    """

    def __init__(self, persist: bool = False, max_retry_time: Optional[float] = None, **kwargs):
        """
        Initialize connection management.

        Args:
            persist: Enable automatic reconnection on disconnect
            max_retry_time: Max total time in ms to keep retrying (None = forever)
            **kwargs: Additional arguments passed to parent class
        """
        super().__init__(**kwargs)
        self._persist = persist
        self._max_retry_time = max_retry_time  # ms; None = retry forever
        self._retry_start_time: Optional[float] = None  # when first failure occurred; used to enforce max_retry_time
        self._current_reconnect_delay: float = 0.25  # seconds until next retry; doubled each failure, capped at 2.5s
        self._manual_disconnect = False  # True only after user calls disconnect(); stops on_disconnected from scheduling reconnect
        self._reconnect_task: Optional[asyncio.Task] = None  # task that sleeps then calls _attempt_connection
        self._did_notify_connected = False  # True after we called on_connected; gates whether we invoke user on_disconnected
        self._connect_result: Optional[ConnectResult] = None  # stored on successful connect

    async def on_connected(self, connection_info: Optional[str] = None) -> None:
        """Handle connection established event."""
        self._manual_disconnect = False
        self._did_notify_connected = True
        self._current_reconnect_delay = 0.25
        self._retry_start_time = None
        await super().on_connected(connection_info)

    async def on_disconnected(self, reason: Optional[str] = None, has_error: bool = False) -> None:
        """
        Handle disconnection event.

        Clears stored ConnectResult, notifies user if we had connected,
        and schedules reconnection in persist mode.
        """
        self._transport = None
        self._connect_result = None

        if self._did_notify_connected:
            self._did_notify_connected = False
            await super().on_disconnected(reason, has_error)

        if self._persist and not self._manual_disconnect:
            await self._schedule_reconnect()

    # =========================================================================
    # INTERNAL CONNECTION PRIMITIVES
    # =========================================================================

    async def _internal_connect(self, timeout: Optional[float] = None) -> Dict[str, Any]:
        """
        Create transport if needed, connect, send auth, and notify on_connected.
        Returns the auth response body (ConnectResult fields). Raises on failure.
        """
        if self._transport is None:
            self._transport = TransportWebSocket(uri=self._uri, auth=self._apikey)
            self._bind_transport(self._transport)

        return await DAPClient.connect(self, timeout)

    async def _internal_disconnect(self, reason: Optional[str] = None, has_error: bool = False) -> None:
        """Close and clean up the transport. Transport invokes on_disconnected when it closes."""
        if self._transport is None:
            return
        await self._transport.disconnect(reason, has_error)

    async def _attempt_connection(self, timeout: Optional[float] = None) -> Optional[ConnectResult]:
        """
        Try _internal_connect once; on auth error stop; on other error reschedule with backoff.
        Used by persist-mode connect() and by the reconnect task.
        Returns ConnectResult on success, None on failure.
        """
        try:
            body = await self._internal_connect(timeout)
            connect_result: ConnectResult = body  # type: ignore[assignment]
            self._connect_result = connect_result
            # Persist mode: store userToken so reconnects use the durable rr_ key
            if connect_result.get('userToken'):
                self._apikey = connect_result['userToken']
            self._reconnect_task = None
            self._debug_message('Connection successful')
            return connect_result
        except AuthenticationException as e:
            self._debug_message(f'Connection failed (auth): {e}')
            await self.on_connect_error(e)
            return None
        except Exception as e:
            self._debug_message(f'Connection failed: {e}')
            await self.on_connect_error(e)

            if self._retry_start_time is None:
                self._retry_start_time = time.monotonic()

            if self._max_retry_time is not None:
                if time.monotonic() - self._retry_start_time >= self._max_retry_time / 1000.0:
                    return None

            self._current_reconnect_delay = min(self._current_reconnect_delay * 2, 2.5)
            await self._schedule_reconnect()
            return None

    async def _schedule_reconnect(self) -> None:
        """Schedule a reconnection attempt with exponential backoff."""
        if self._reconnect_task and not self._reconnect_task.done():
            self._reconnect_task.cancel()

        if self._max_retry_time is not None and self._retry_start_time is not None:
            if time.monotonic() - self._retry_start_time >= self._max_retry_time / 1000.0:
                await self.on_connect_error(Exception('Max retry time exceeded'))
                return

        self._debug_message(f'Scheduling reconnection in {self._current_reconnect_delay}s')
        self._reconnect_task = asyncio.create_task(self._attempt_reconnect())

    async def _attempt_reconnect(self) -> None:
        """Sleep then call _attempt_connection (used by scheduled reconnect)."""
        await asyncio.sleep(self._current_reconnect_delay)
        if self._persist and not self._manual_disconnect:
            self._debug_message('Attempting to reconnect...')
            await self._attempt_connection()

    # =========================================================================
    # PUBLIC API
    # =========================================================================

    async def connect(
        self,
        credential: Optional[str] = None,
        *,
        timeout: Optional[float] = None,
    ) -> ConnectResult:
        """
        Connect to the RocketRide server and authenticate in a single call.

        Sends the credential as the first DAP message and returns the full
        ConnectResult (user identity + organizations + teams) on success.

        If `credential` is omitted, falls back to: the `auth` passed at construction
        time, then the `ROCKETRIDE_APIKEY` environment variable.

        In persist mode, enables automatic reconnection. After the first successful
        connect the stored `userToken` is replayed automatically on reconnect.

        Args:
            credential: API key / Zitadel access_token / rr_ user token.
                        Falls back to construction-time auth, then ROCKETRIDE_APIKEY env var.
            timeout: Optional overall timeout in ms covering the WebSocket handshake
                     and auth request.

        Returns:
            ConnectResult with user identity, organizations, and teams.
        """
        resolved = credential or self._apikey or os.environ.get('ROCKETRIDE_APIKEY', '')
        self._set_auth(resolved)

        self._manual_disconnect = False
        self._current_reconnect_delay = 0.25
        self._retry_start_time = None

        if self.is_connected():
            await self._internal_disconnect()

        result: Optional[ConnectResult] = None
        if self._persist:
            if self._reconnect_task and not self._reconnect_task.done():
                self._reconnect_task.cancel()
                self._reconnect_task = None
            result = await self._attempt_connection(timeout)
        else:
            body = await self._internal_connect(timeout)
            result = body  # type: ignore[assignment]
            self._connect_result = result
            if result and result.get('userToken'):
                self._apikey = result['userToken']

        return result or {}  # type: ignore[return-value]

    def get_account_info(self) -> Optional[ConnectResult]:
        """
        Return the ConnectResult from the last successful connect().
        Returns None if not connected or not yet authenticated.
        """
        return self._connect_result

    async def disconnect(self) -> None:
        """
        Disconnect from the RocketRide server and stop automatic reconnection.

        Should be called when finished with the client to clean up resources.
        """
        self._manual_disconnect = True

        if self._reconnect_task and not self._reconnect_task.done():
            self._reconnect_task.cancel()
            self._reconnect_task = None

        if self._transport is not None and self.is_connected():
            await self._internal_disconnect()

    # =========================================================================
    # HELPERS
    # =========================================================================

    def get_connection_info(self) -> dict:
        """Return current connection state and URI."""
        return {
            'connected': self.is_connected(),
            'transport': 'WebSocket',
            'uri': getattr(self, '_uri', ''),
        }

    def get_apikey(self) -> Optional[str]:
        """Return the API key in use. For debugging only."""
        return getattr(self, '_apikey', None)

    @staticmethod
    def normalize_uri(uri: str) -> str:
        """Normalize a user-provided URI into a fully-formed HTTP/HTTPS URL."""
        if uri and '://' not in uri:
            uri = f'{CONST_DEFAULT_WEB_PROTOCOL}{uri}'

        parsed = urllib.parse.urlparse(uri)

        if not parsed.port and 'rocketride.ai' not in (parsed.hostname or ''):
            hostname = parsed.hostname
            if not hostname:
                raise ValueError(f"Invalid URI '{uri}': missing hostname")
            parsed = parsed._replace(netloc=f'{hostname}:{CONST_DEFAULT_WEB_PORT}')

        return parsed.geturl()

    @staticmethod
    def _get_websocket_uri(uri: str) -> str:
        """Normalize a user-provided URI into a fully-formed WebSocket address."""
        normalized = ConnectionMixin.normalize_uri(uri)
        parsed = urllib.parse.urlparse(normalized)

        ws_scheme = 'wss' if parsed.scheme in ('https', 'wss') else 'ws'
        ws_uri = parsed._replace(scheme=ws_scheme)
        return f'{ws_uri.geturl()}/task/service'

    def _set_uri(self, uri: str) -> None:
        """Update the server URI (internal)."""
        self._uri = self._get_websocket_uri(uri)

    def _set_auth(self, auth: str) -> None:
        """Update the authentication credential (internal)."""
        self._apikey = auth

    def set_env(self, env: Dict[str, str]) -> None:
        """Update the environment variables used for pipeline substitution."""
        self._env = dict(env)

    async def request(self, request: Dict[str, Any], timeout: Optional[float] = None) -> Dict[str, Any]:
        """Send a request to the RocketRide server."""
        return await super().request(request, timeout=timeout)

File: packages/chat-widget/src/connection.ts
/**
 * MIT License
 *
 * Copyright (c) 2026 Aparavi Software AG
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
 * Protocol layer for the chat widget — a UI-free wrapper around the workspace
 * 'rocketride' TypeScript SDK.
 *
 * AUTH MODEL (security-critical): the widget authenticates with the pipeline's
 * PUBLIC auth key only — the same `{host}/chat?auth={public_auth}` credential
 * the chat source node publishes for end users. NEVER configure the widget
 * with an engine API key or any private token: the auth value is embedded in
 * the host page and is visible to every visitor. The public key both
 * authenticates the WebSocket connection and addresses the pipeline
 * (`client.chat({ token })`), mirroring apps/chat-ui.
 *
 * TRANSPORT SECURITY (security-critical): the SDK maps a non-TLS engine URI to
 * a plain `ws:` socket (see `RocketRideClient._getWebsocketUri`), which would
 * put the auth key and the whole conversation on the wire in cleartext. This
 * module therefore refuses cleartext for anything but loopback development
 * hosts — see {@link assertSecureEngineUrl}.
 *
 * Reconnection is owned by the SDK (`persist: true`): after a drop it retries
 * with linear backoff and never gives up; this class only translates the SDK
 * callbacks into UI-friendly state changes.
 *
 * @module connection
 */

import { ConnectionException, PIPELINE_RESULT, Question, QuestionType, RocketRideClient, RocketRideClientConfig } from 'rocketride';
import { ChatClientFactory, ChatClientLike, ChatHistoryItem, ConnectionState } from './types';

/** Maximum number of prior messages replayed to the pipeline for context (mirrors apps/chat-ui). */
export const HISTORY_LIMIT = 6;

/** Rejection message handed to an {@link WidgetConnection.ask} that {@link WidgetConnection.disconnect} closed the transport under. */
export const ASK_ABANDONED_MESSAGE = 'The connection was closed before the pipeline answered — please try again.';

/** Options for {@link WidgetConnection}. */
export interface WidgetConnectionOptions {
	/**
	 * Engine URL, e.g. `https://engine.example.com` (normalised by the SDK).
	 * Must be TLS (`https:`/`wss:`) unless it points at a loopback host — see
	 * {@link assertSecureEngineUrl}.
	 */
	engineUrl: string;
	/**
	 * The pipeline's PUBLIC auth key — the `?auth=` value from the chat source
	 * node's published link. Never an engine API key or private token.
	 */
	auth: string;
	/** Notified whenever the connection state changes. `detail` carries the error/disconnect reason. */
	onStateChange?: (state: ConnectionState, detail?: string) => void;
	/**
	 * Notified with live pipeline status ('thinking') lines while a question is
	 * processed. Used only for {@link WidgetConnection.ask} calls that pass no
	 * per-request status callback: status belongs to one request, and a caller
	 * with more than one in flight has to receive it per request to know whose
	 * it is.
	 */
	onStatus?: (text: string) => void;
	/** Client factory override for headless unit tests. Defaults to the real SDK client. */
	createClient?: ChatClientFactory;
}

/**
 * Extracts displayable answer texts from a pipeline result.
 *
 * `result_types` maps dynamic field names to data types; fields typed 'text'
 * or 'answers' hold displayable strings (either directly, as arrays, or as
 * `{ answer: string }` objects). Mirrors apps/chat-ui/src/utils/pipelineUtils.
 *
 * @param result - Pipeline result returned by `client.chat()` (narrowed defensively)
 * @returns Answer strings in result order (empty when the pipeline returned none)
 */
export function extractAnswerTexts(result: PIPELINE_RESULT | unknown): string[] {
	const texts: string[] = [];
	if (!result || typeof result !== 'object') {
		return texts;
	}
	const pipelineResult = result as PIPELINE_RESULT;
	if (!pipelineResult.result_types || typeof pipelineResult.result_types !== 'object') {
		return texts;
	}

	for (const [fieldName, fieldType] of Object.entries(pipelineResult.result_types)) {
		if (fieldType !== 'text' && fieldType !== 'answers') {
			continue;
		}
		const fieldData: unknown = pipelineResult[fieldName];
		if (Array.isArray(fieldData)) {
			for (const item of fieldData) {
				if (typeof item === 'string' && item.trim()) {
					texts.push(item);
				}
			}
		} else if (typeof fieldData === 'string' && fieldData.trim()) {
			texts.push(fieldData);
		} else if (typeof fieldData === 'object' && fieldData !== null && typeof (fieldData as Record<string, unknown>).answer === 'string') {
			// Answer objects arrive as { answer: string, expectJson: bool }.
			const answer = ((fieldData as Record<string, unknown>).answer as string).trim();
			if (answer) {
				texts.push(answer);
			}
		}
	}

	return texts;
}

/** Schemes the widget accepts in `engineUrl`; anything else is a configuration error. */
const ACCEPTED_SCHEMES = new Set(['https:', 'wss:', 'http:', 'ws:']);

/** Schemes that encrypt the transport; the rest are cleartext. */
const SECURE_SCHEMES = new Set(['https:', 'wss:']);

/**
 * True for hostnames that never leave the machine, so cleartext is acceptable
 * there: `localhost` (and any `*.localhost` subdomain, which RFC 6761 reserves
 * for loopback), the whole `127.0.0.0/8` range, and the IPv6 loopback.
 *
 * @param hostname - `URL.hostname` (IPv6 literals arrive without brackets)
 */
function isLoopbackHost(hostname: string): boolean {
	// URL.hostname keeps the brackets off IPv6 literals, but be defensive.
	const host = hostname.toLowerCase().replace(/^\[|\]$/g, '');
	return host === 'localhost' || host.endsWith('.localhost') || host === '::1' || host === '0:0:0:0:0:0:0:1' || /^127\.\d{1,3}\.\d{1,3}\.\d{1,3}$/.test(host);
}

/**
 * Rejects an engine URL whose transport would carry the auth key and the
 * conversation in cleartext.
 *
 * The SDK downgrades silently: `RocketRideClient.normalizeUri` prefixes a
 * scheme-less URI with `http://`, and `_getWebsocketUri` turns anything that is
 * not `https:`/`wss:` into a plain `ws:` socket. A widget is embedded in
 * arbitrary third-party pages, so that downgrade has to be caught here rather
 * than left to the operator: `https:`/`wss:` is required for every host except
 * loopback, where cleartext stays allowed for local development (the demo page
 * and `http://localhost:5565` keep working unchanged).
 *
 * @param engineUrl - The configured engine URL, exactly as the embedder wrote it
 * @throws Error when the URL is unparseable, uses an unsupported scheme, or is
 *   cleartext against a non-loopback host
 */
export function assertSecureEngineUrl(engineUrl: string): void {
	const raw = engineUrl.trim();
	// Mirror the SDK's scheme defaulting so a bare 'engine.example:5565' is
	// judged as the http:// URL the SDK will actually build from it.
	const candidate = /^[a-zA-Z][a-zA-Z\d+\-.]*:\/\//.test(raw) ? raw : `http://${raw}`;

	let url: URL;
	try {
		url = new URL(candidate);
	} catch {
		throw new Error(`WidgetConnection: engineUrl is not a valid URL (${engineUrl}).`);
	}

	if (!ACCEPTED_SCHEMES.has(url.protocol)) {
		throw new Error(`WidgetConnection: unsupported engineUrl scheme "${url.protocol}" — use https:// or wss:// (http:// and ws:// are accepted for loopback hosts only).`);
	}
	if (SECURE_SCHEMES.has(url.protocol) || isLoopbackHost(url.hostname)) {
		return;
	}

	throw new Error(`WidgetConnection: refusing to connect to "${engineUrl}" over cleartext — the auth key and every message would cross the network unencrypted. Use https:// or wss:// (terminate TLS in front of the engine); plain http:// or ws:// is allowed only for loopback hosts (localhost, 127.0.0.1, ::1) during local development.`);
}

/**
 * Manages one persistent widget connection to a RocketRide engine.
 *
 * UI-free by design: all UI concerns are delivered through the callbacks in
 * {@link WidgetConnectionOptions}, so this class is unit-testable headlessly
 * with an injected {@link ChatClientFactory}.
 */
export class WidgetConnection {
	private readonly _options: WidgetConnectionOptions;
	private _client: ChatClientLike | null = null;
	private _state: ConnectionState = 'idle';
	private _manualDisconnect = false;
	/**
	 * Rejectors for every {@link ask} still awaiting an answer. `disconnect()`
	 * closes the socket underneath the SDK's `chat()` promise, which may then
	 * never settle; rejecting through these gives the caller a definite outcome
	 * instead of stranding it — and, in the widget, a composer that would stay
	 * disabled behind a request that can no longer arrive.
	 */
	private readonly _pendingAsks = new Set<(error: Error) => void>();

	/** Stores the connection options; no client is created until `connect()` is called. */
	constructor(options: WidgetConnectionOptions) {
		this._options = options;
	}

	/** Current connection state. */
	get state(): ConnectionState {
		return this._state;
	}

	/** True once the transport is connected and authenticated. */
	isConnected(): boolean {
		return this._state === 'connected' && this._client !== null && this._client.isConnected();
	}

	/**
	 * Opens the persistent connection. Resolves once the first attempt
	 * completes; with `persist` the SDK keeps retrying on failure/drop, so a
	 * resolved promise does not imply the state is 'connected' — observe
	 * `onStateChange`. Idempotent while a client exists.
	 *
	 * @throws Error when engineUrl or auth is missing
	 */
	async connect(): Promise<void> {
		if (this._client) {
			return;
		}
		if (!this._options.engineUrl || !this._options.auth) {
			throw new Error("WidgetConnection requires both engineUrl and auth (the pipeline's public auth key).");
		}
		// Security gate: never hand a credentialed connection to a cleartext
		// transport. Throws before any client exists, so nothing is sent.
		assertSecureEngineUrl(this._options.engineUrl);

		this._setState('connecting');
		this._manualDisconnect = false;

		const config: RocketRideClientConfig = {
			// PUBLIC auth key only — see the module doc comment.
			auth: this._options.auth,
			uri: this._options.engineUrl,
			// The SDK owns reconnection: linear backoff, retries forever.
			persist: true,
			// Never fall back to ambient env credentials (e.g. ROCKETRIDE_APIKEY).
			env: {},
			module: 'chat-widget',
			clientName: 'RocketRide Chat Widget',
			onConnected: async () => {
				this._setState('connected');
			},
			onDisconnected: async (reason?: string, hasError?: boolean) => {
				if (this._manualDisconnect) {
					return;
				}
				// The SDK immediately schedules a reconnect; reflect an error state
				// for server-side rejections, otherwise show 'connecting'.
				this._setState(hasError ? 'error' : 'connecting', reason || 'Connection lost');
			},
			onConnectError: (error: ConnectionException) => {
				this._setState('error', error.message);
			},
		};

		const factory: ChatClientFactory = this._options.createClient ?? ((clientConfig) => new RocketRideClient(clientConfig));
		this._client = factory(config);
		await this._client.connect();
	}

	/** Closes the connection, abandons anything still in flight on it, and disables automatic reconnection. */
	async disconnect(): Promise<void> {
		this._manualDisconnect = true;
		// Settle in-flight asks first, and synchronously: a caller that tears
		// this connection down may rely on no request still being pending on it
		// by the time `disconnect()` returns its promise.
		this._abandonPendingAsks();
		const client = this._client;
		this._client = null;
		if (client) {
			await client.disconnect();
		}
		this._setState('idle');
	}

	/** Force-retries the connection immediately (error-state Retry button). */
	async retry(): Promise<void> {
		await this.disconnect();
		await this.connect();
	}

	/**
	 * Sends one question through the pipeline and resolves with the answers.
	 *
	 * @param text - The user's question
	 * @param history - Prior conversation turns for context; only the last
	 *   {@link HISTORY_LIMIT} are sent (UI-only entries must be pre-filtered)
	 * @param onStatus - Receives the live status ('thinking') lines of THIS
	 *   request. Status is a property of a request, not of the connection: a
	 *   caller that abandons an `ask()` (through `disconnect()`, or by clearing
	 *   the transcript it belonged to) can start the next one on the same
	 *   connection while the old SDK `chat()` is still streaming, and only a
	 *   per-request callback lets it tell the two apart. Falls back to the
	 *   connection-level {@link WidgetConnectionOptions.onStatus} when omitted.
	 * @returns Answer texts extracted from the pipeline result
	 * @throws Error when the connection has not been opened or the pipeline fails
	 */
	async ask(text: string, history: ChatHistoryItem[] = [], onStatus?: (text: string) => void): Promise<string[]> {
		const client = this._client;
		if (!client) {
			throw new Error('Not connected to RocketRide — call connect() first.');
		}

		const question = new Question({ type: QuestionType.PROMPT, expectJson: false });
		question.addQuestion(text);
		for (const item of history.slice(-HISTORY_LIMIT)) {
			question.addHistory({ role: item.role, content: item.content });
		}

		// Bind this request's status sink once, so every line the SDK streams is
		// attributable to the ask() that produced it.
		const forwardStatus = onStatus ?? this._options.onStatus;

		const chat = client.chat({
			// The pipeline is addressed with the same public auth key.
			token: this._options.auth,
			question,
			onSSE: async (_type: string, data: Record<string, unknown>) => {
				const message = data.message;
				if (typeof message === 'string' && message) {
					forwardStatus?.(message);
				}
			},
		});

		// Race the SDK call against an abandon signal so a `disconnect()` under
		// it (element removed, engine/auth swap, Retry) always settles this
		// promise instead of leaving the caller waiting on a dead socket.
		let abandon!: (error: Error) => void;
		const abandoned = new Promise<never>((_resolve, reject) => {
			abandon = reject;
		});
		this._pendingAsks.add(abandon);

		try {
			return extractAnswerTexts(await Promise.race([chat, abandoned]));
		} finally {
			this._pendingAsks.delete(abandon);
		}
	}

	/**
	 * Rejects every in-flight {@link ask} with {@link ASK_ABANDONED_MESSAGE}.
	 * The underlying SDK call is left to settle (or not) on its own; its result
	 * is discarded by the {@link Promise.race} in `ask()`.
	 */
	private _abandonPendingAsks(): void {
		if (this._pendingAsks.size === 0) {
			return;
		}
		const pending = Array.from(this._pendingAsks);
		this._pendingAsks.clear();
		for (const abandon of pending) {
			abandon(new Error(ASK_ABANDONED_MESSAGE));
		}
	}

	/** Records the new state and notifies the `onStateChange` callback. */
	private _setState(state: ConnectionState, detail?: string): void {
		this._state = state;
		this._options.onStateChange?.(state, detail);
	}
}
